engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/execution/ReentrantLocking.scala

```diff
@@ -126,4 +132,57 @@ class ReentrantLocking(logger: TruffleLogger) extends Locking {
     now2
   }
 
+  /** @inheritdoc */
+  override def getUniqueTaskIndex(): Int = {
+    orderLock.lock()
+    try {
+      taskIndex.enqueue()
+    } finally {
+      orderLock.unlock()
+    }
+  }
+
+  /** @inheritdoc */
+  override def waitOnTaskIndex(index: Int): Unit = {
+    orderLock.lock()
+    try {
+      val current = taskIndex.getCurrent()
+      while (current != index) {
+        indexAdvancedCondition.await()
+      }
+    } finally {
+      orderLock.unlock()
+    }
+  }
+
+  /** @inheritdoc */
+  override def advanceTaskIndex(): Unit = {
+    orderLock.lock()
+    try {
+      taskIndex.dequeue()
+      indexAdvancedCondition.signalAll()
+    } finally {
+      orderLock.unlock()
+    }
+  }
+
+  private class OrderedTaskIndex {
+    private var current = 0
+    private var max     = 0
+
+    def dequeue(): Unit = {
+      current = current + 1
+    }
+
+    def getCurrent(): Int = {
+      current
+    }
+
+    def enqueue(): Int = {
+      val idx = max
+      max = max + 1
+      assert(max > current)
+      idx
+    }
```

Comment:
What happens if the index overflows? IMO assertion is not enough - we should have a check that runs without asserts enabled as well.

We may also change this to `Long` to give it more space before it overflows